File: need/Cuber.py
# from my_module import Cube
from typing import List
# import numba as nb
import open3d as o3d
import numpy as np


# @nb.jitclass
# class Point:
#     def __init__(self, x, y):
#         self.x = x
#         self.y = y


# @nb.jitclass
class Cube:
    def __init__(self, p1, p2):
        self.p1 = p1
        self.p2 = p2
        self.size = np.abs(p1 - p2)

    def is_valid(self, max_size):
        return np.all(self.size <= max_size)

    def __eq__(self, other):
        return np.all(self.p1 == other.p1) and np.all(self.p2 == other.p2)


def find_closest_points(points, pt):
    # 计算每个点与目标点之间的欧几里得距离
    dists = np.sqrt(np.sum((points - pt)**2, axis=1))

    # 找到距离最近的两个点的索引
    closest_idxs = np.argsort(dists)[:2]

    return closest_idxs


# @nb.njit("List[Cube](np.ndarray, float)")
# @nb.jit(nopython=True)
def create_cube(points: np.ndarray, max_size: float) -> List[Cube]:
    cubes = np.empty((len(points), len(points)), dtype=Cube)  # 创建空的NumPy数组
    cube_count = 0  # 立方体计数器

    for i in range(len(points)-1):
        for j in range(i+1, len(points)):
            p1 = points[i]
            p2 = points[j]
            cube = Cube(p1, p2)

            if cube.is_valid(max_size) and cube not in cubes:
                cubes[cube_count] = cube
                cube_count += 1

    return cubes[:cube_count].tolist()  # 返回有效的立方体列表


# @nb.njit
def Sampledata(pcd: o3d.geometry.PointCloud):  # （Open3D点云对象）
    # pcd = o3d.io.read_point_cloud("point_cloud.ply")
    points = np.asarray(pcd.points)
    max_size = 5

    cubes = create_cube(points, max_size)
    # 创建可视化
    vis = o3d.visualization.Visualizer()
    vis.create_window()

    # 将正方体添加到可视化中
    for cube in cubes:
        box = o3d.geometry.OrientedBoundingBox(cube.p1, cube.p2)
        vis.add_geometry(box)

    # 设置相机视角
    vis.get_view_control().set_front([0, 0, -1])
    vis.run()
    vis.destroy_window()

File: Cuber.cpp
#include <iostream>
#include <vector>
#include <cmath>

struct Point {
    double x;
    double y;
};

class Cube {
public:
    Point p1;
    Point p2;
    Point size;

    Cube(const Point& p1, const Point& p2) : p1(p1), p2(p2) {
        size.x = std::abs(p1.x - p2.x);
        size.y = std::abs(p1.y - p2.y);
    }

    bool isValid(double max_size) {
        return (size.x <= max_size) && (size.y <= max_size);
    }

    bool operator==(const Cube& other) {
        return (p1.x == other.p1.x) && (p1.y == other.p1.y) && (p2.x == other.p2.x) && (p2.y == other.p2.y);
    }
};

std::vector<Cube> createCube(const std::vector<Point>& points, double max_size) {
    std::vector<Cube> cubes;
    int cube_count = 0;

    for (int i = 0; i < points.size() - 1; i++) {
        for (int j = i + 1; j < points.size(); j++) {
            const Point& p1 = points[i];
            const Point& p2 = points[j];
            Cube cube(p1, p2);

            if (cube.isValid(max_size)) {
                bool exists = false;
                for (const auto& existing_cube : cubes) {
                    if (cube == existing_cube) {
                        exists = true;
                        break;
                    }
                }
                if (!exists) {
                    cubes.push_back(cube);
                    cube_count++;
                }
            }
        }
    }

    return std::vector<Cube>(cubes.begin(), cubes.begin() + cube_count);
}
// #include <iostream>
// #include <Open3D/Visualization/Visualizer.h>
// #include <Open3D/Geometry/OrientedBoundingBox.h>
// void sampleData(const std::vector<Cube>& cubes) {
//     // 创建可视化对象
//     o3d::visualization::Visualizer vis;
//     vis.CreateVisualizerWindow("Open3D Viewer", 1920, 1080);

//     // 将正方体添加到可视化中
//     for (const auto& cube : cubes) {
//         const o3d::geometry::OrientedBoundingBox box = cube.GetBoundingBox();
//         vis.AddGeometry(box);
//     }

//     // 设置相机视角
//     vis.GetViewControl().SetFront({0, 0, -1});

//     // 运行可视化
//     vis.Run();

//     // 销毁窗口
//     vis.DestroyVisualizerWindow();
// }


#include <iostream>
#include <pcl/visualization/pcl_visualizer.h>
#include <pcl/common/common.h>
#include <pcl/common/transforms.h>

typedef pcl::PointXYZ PointT;

void sampleData(const std::vector<Cube>& cubes) {
    // 创建可视化对象
    pcl::visualization::PCLVisualizer visualizer("PCL Viewer");

    // 设置背景颜色和坐标轴
    visualizer.setBackgroundColor(0, 0, 0);
    visualizer.addCoordinateSystem(1.0);

    // 将正方体添加到可视化中
    for (const auto& cube : cubes) {
        const Eigen::Vector3f center = cube.GetCenter();
        const float length = cube.GetLength();

        // 构建正方体的八个顶点
        pcl::PointCloud<PointT>::Ptr cubeCloud(new pcl::PointCloud<PointT>());
        cubeCloud->points.resize(8);
        cubeCloud->points[0].getVector3fMap() = center + Eigen::Vector3f(-length / 2, -length / 2, -length / 2);
        cubeCloud->points[1].getVector3fMap() = center + Eigen::Vector3f(length / 2, -length / 2, -length / 2);
        // ... 添加剩余的点

        // 创建可视化对象并添加到可视化中
        pcl::visualization::PointCloudColorHandlerCustom<PointT> color_handler(cubeCloud, 255, 0, 0);  // 红色
        visualizer.addPointCloud(cubeCloud, color_handler);
    }

    // 设置相机视角
    visualizer.setCameraPosition(0, 0, -5, 0, -1, 0, 0);

    // 运行可视化
    while (!visualizer.wasStopped()) {
        visualizer.spinOnce(100);
    }
}



int main() {
    // 从文件或其他方式获取点数据并存储在std::vector<Point>中
    std::vector<Point> points;

    // 调用sampleData函数进行处理和可视化
    sampleData(points);

    return 0;
}
